lib/wibox/template.lua

```diff
@@ -0,0 +1,439 @@
+---------------------------------------------------------------------------
+-- An abstract widget that handles a preset of concrete widget.
+--
+-- The `wibox.template` widget is an abstraction layer that contains a
+-- concrete widget definition. The template widget can be used to build widgets
+-- that the user can customize at their will, thanks to the template mechanism.
+--
+-- Common usage examples
+-- =====================
+--
+-- A basic implementation of the template widget needs a widget definition and
+-- a callback to manage widget updates.
+--
+-- The `:update()` can be called later to update the widget template. Arguments
+-- can be provided to the `:update()` method, it will be forwarded to the
+-- `update_callback` function.
+--
+--@DOC_wibox_widget_template_basic_textbox_EXAMPLE@
+--
+-- Usage in libraries
+-- ==================
+--
+-- (this part is to be completed according to awful.widget comming PRs to implement the widget_template usage)
+--
+-- This widget was designed to be used as a standard way to offer customization
+-- over concrete widget implementation. We use the `wibox.template` as a
+-- base to implement widgets from the `awful.widget` library. This way, it is
+-- easy for the final user to customize the standard widget offered by awesome!
+--
+-- It is possible to use the template widget as a base for a custom 3rd party
+-- widget module to offer more customization to the final user.
+--
+-- Here is an example of implementation for a custom widget inheriting from
+-- `wibox.template` :
+--
+-- The module definition should include a default widget and a builder function
+-- that can build the widget with either, the user values or the default.
+--
+--@DOC_wibox_widget_template_concrete_implementation_module_EXAMPLE@
+--
+-- On the user side, it only requires to call the builder function to get an
+-- instance of the widget.
+--
+--@DOC_wibox_widget_template_concrete_implementation_user_EXAMPLE@
+--
+-- @author Aire-One
+-- @copyright 2021 Aire-One <aireone@aireone.xyz>
+--
+-- @widgetmod wibox.template
+-- @supermodule wibox.widget.base
+---------------------------------------------------------------------------
+
+local wbase = require("wibox.widget.base")
+local gtable = require("gears.table")
+local gtimer = require("gears.timer")
+
+local template = {
+    mt = {},
+    queued_updates = {},
```

Comment:
Since the keys in this table will be widgets, it should probably be defined with weak references on those keys.